### DISCUSSION TITLE
Removal of widgets, text is lost

### DISCUSSION THREAD

**Discussion by AndresEV22** [pymupdf/PyMuPDF#2574] [Q&A] [answered]
When I use the method to delete all widgets from a PDF I lose all the text inside it, or when I use the widget.field_flags = fitz.PDF_FIELD_IS_READ_ONLY if I open the fields again they are editable.

## To Reproduce (mandatory)
On read only fields, i save with this method:
for widget in widgets:
                ... logic to change the field value
                widget.field_flags = fitz.PDF_FIELD_IS_READ_ONLY
                widget.update()

and to delete it: 
for widget in widgets:
                widget.field_value = fields[widget.field_name]  # type: ignore
                widget.field_flags = fitz.PDF_FIELD_IS_READ_ONLY
                widget.update()
                page.delete_widget(widget)

## Expected behavior (optional)
I expect text within a widget to be saved and the widget disappears or with the field flag when I reopen the PDF it's fields shouldn't be editable.

## Your configuration (mandatory)
 - Windows 10 and arch linux
 - Python 3.11
 - PyMuPDF 1.22.5, installation method  by pyproyect.toml.

**Comment by JorjMcKie** [comment]
Deleting a widget, in general cannot be prevented by setting a field read-only.
That flag only prevents the widget being **_changed_**. You can still delete it independently from this.

Otherwise, I don't understand:
If you delete a widget, then it will be gone - together with everything associated.

**Comment by AndresEV22** [comment]
The thing I want to do is a common PDF operation called flattening. What this does is combine all text into one layer, removing the widgets and making it more hard to edit so we can protect the information that was filled out in the form. I know pdftk does it and also applications like Adobe Acrobat Reader.

This is mostly necessary to store the PDF file securely and increasing security on the document so it cannot be filled out later again.

Is this possible in pymupdf?

this is an exact example of flattening operation: https://pdfpro.com/blog/guides/how-to-flatten-a-pdf/

**Comment by JorjMcKie** [comment, marked as answer]
In contrast to what the referenced article implies, the term "flattening" is **_not_** clearly defined.
Some people understand this as creating an image-only version for each page.
Otherwise, there is `doc.convert_to_pdf()` which also converts widgets and annotations to normal, standard PDF content. It also removes PDF Optional Content layers.

**Comment by AndresEV22** [comment]
okay, thank you very much for the help
